cmd/argocd/commands/completion.go

```diff
@@ -0,0 +1,211 @@
+package commands
```

Comment:
Can I please ask - is this generated?

Reply:
https://github.com/spf13/cobra/blob/master/bash_completions.md

Reply:
I think this can be auto-generated, but it is not?

Reply:
Yeah, I tried that and I got an error and could not figure out how to recover.

```
Edwins-MacBook-Pro:argocd edwin$ cobra add completion
Error: open /Users/edwin/go/src/github.com/argoproj/argo-cd/cmd/argocd/cmd/completion.go: no such file or directory
Edwins-MacBook-Pro:argocd edwin$ cobra --help
Cobra is a CLI library for Go that empowers applications.
This application is a tool to generate the needed files
to quickly create a Cobra application.

Usage:
  cobra [command]

Available Commands:
  add         Add a command to a Cobra Application
  help        Help about any command
  init        Initialize a Cobra Application

Flags:
  -a, --author string    author name for copyright attribution (default "YOUR NAME")
      --config string    config file (default is $HOME/.cobra.yaml)
  -h, --help             help for cobra
  -l, --license string   name of license for the project
      --viper            use Viper for configuration (default true)

Use "cobra [command] --help" for more information about a command.
```

I'm going to guess this is because the argocd is leveraging older cobra conventions, though I did not do the code archaeology to confirm.

So, what I did was copy the completion.go from the Argo Project. The custom completion bash code still has to be hand-written.  Cobra will make all the static completions work automatically so it's a lot less work.

Reply:
ah - I get it. I'd be strongly in favor of generating this, as I know that we'll always forget to fix this. I might try myself shortly.

Reply:
Well, to be clear, the code that's generated is boilerplate.  It's essentially static.  The only thing that's custom are the little scripts to do the automated lookups of things like app history, app names, repos, projects, namespaces, and roles.  The only maintenance would be if the CLI by with the lookups are done is changed, which I think is unlikely.

Reply:
In other words, I basically did the same thing the auto-generator would do.  And unless we went to a new major version of cobra, I don't think the generator will ever have to be re-run.  So it should be no/low maintenance.